packages/dev-server-hmr/src/babel/babelPluginWcHmr.js

```diff
@@ -0,0 +1,241 @@
+ * @param {NodePath<any>} nodePath
+ * @returns {NodePath<any> | undefined}
+ */
+function walkClassMixins(nodePath) {
+  let el = nodePath.node.superClass;
+  // walk possible mixin functions
+  while (el && t.isCallExpression(el)) {
+    [el] = el.arguments;
+  }
+  return el;
+}
+
+/**
+ * @param {BaseClass[]} baseClasses
+ * @param {string} importSpecifier
+ * @param {string} filename
+ * @param {string} rootDir
+ * @return {{ baseClass: BaseClass, i: number }[]}
+ */
+function getBaseClassesForImport(baseClasses, importSpecifier, filename, rootDir) {
+  return baseClasses
+    .map((baseClass, i) => ({ baseClass, i }))
+    .filter(({ baseClass }) => {
+      if (!baseClass.import) {
+        return false;
+      }
+
+      if (importSpecifier.startsWith('./') || importSpecifier.startsWith('/')) {
+        // this is a non-bare import
+        const partialImportFilePath = importSpecifier.split('/').join(path.sep);
+        const joinBase = importSpecifier.startsWith('/') ? rootDir : path.dirname(filename);
+        const importFilePath = path.join(joinBase, partialImportFilePath);
+        return importFilePath === baseClass.import;
+      }
+
+      // this is a bare import
+      if (baseClass.import === importSpecifier) {
+        return true;
+      }
+
+      if (!path.extname(baseClass.import) && `${baseClass.import}.js` === importSpecifier) {
+        return true;
+      }
+
+      return false;
+    });
+}
+
+/**
+ * @param {BaseClass[]} baseClasses
+ * @param {(ImportDefaultSpecifier | ImportNamespaceSpecifier | ImportSpecifier)[]} specifiers
+ */
+function getImportedBaseClassNames(baseClasses, specifiers) {
+  /** @type {string[]} */
+  const names = [];
+
+  for (const specifier of specifiers) {
+    if (t.isImportDefaultSpecifier(specifier)) {
+      if (baseClasses.some(cl => cl.name === 'default')) {
+        names.push(specifier.local.name);
+      }
+    } else if (t.isImportSpecifier(specifier)) {
+      const imported = t.isIdentifier(specifier.imported)
+        ? specifier.imported.name
+        : specifier.imported.value;
+      if (baseClasses.some(cl => cl.name === imported)) {
+        names.push(specifier.local.name);
+      }
+    }
+  }
+  return names;
+}
+
+/**
+ * @returns {PluginObj}
+ */
+function babelPluginWcHmr() {
+  return {
+    visitor: {
+      Program(program) {
+        if (!this.filename) throw createError('Missing filename');
+        const resolvedFilename = path.resolve(this.filename);
+        const options = parseOptions(/** @type {BabelPluginWcHmrOptions} */ (this.opts));
+
+        /** @type {string[]} */
+        const baseClassNames = [];
+        /** @type {string[]} */
+        const baseClassPatches = [];
+        let injectedRegister = false;
+
+        // collect all base classes we should match without an import
+        for (const b of options.baseClasses) {
+          if (!b.import) {
+            baseClassNames.push(b.name);
+          }
+        }
+
+        program.traverse({
```

Comment:
aren't you triggering a double-traversal here? presumably whatever calls this file as a whole is already traversing the AST (hence the visitor you're returning). so you should instead keep your state outside this `Program` function so you can move your `ImportDeclaration` etc top level.

Reply:
We have to do two traversals because we first collect all import statements, and then inspect the base classes. Import statements aren't guaranteed to always be on top.

Reply:
no but you can visit imports and class declarations/expressions, then have an exit visitor for `Program`:

```ts
Program: {
  exit(node) {
    // by this point, your import and class visitors already ran
  }
}
```

Reply:
When I'm encountering a class, I already need to know which imports there are in this file.